Inform the users if there is a winner

christophermanahan opened the pull request:
### Description

Augment game with a private win method and provide an intermediate step toward a refactor of board that exposes the columns and diagonals to the game class

christophermanahan committed:
Add columns and diagonals to board

christophermanahan committed:
Add win to game and utilize in run

christophermanahan committed:
Add winner to user input and utilize in game
Intermediate step before refactoring game to use a presenter that utilizes an abstraction where the unstable formatter class is not a dependency of user interface

ssunday commented on lib/board.rb:
```diff
@@ -1,12 +1,26 @@
+  end
+
+  def columns
+    (0..size - 1).inject([]) do |acc, mod|
```
Couldn't this be replaced by map?

ssunday commented on lib/game.rb:
```diff
@@ -6,14 +6,25 @@ def initialize(board:, symbols:, user_interface:)
-    board.put(symbols.next, user_interface.get_input)
-    run
+    user_interface.display(board.rows)
+    if win?([board.rows, board.columns, board.diagonals])
```
I wonder if `[board.rows, board.columns, board.diagonals]` could be extracted into a public method in board so game doesn't need to know these details?

christophermanahan replied:
That is my next refactor! :)

ssunday replied:
Woot!

christophermanahan replied:
Hopefully with some inverted control...

christophermanahan replied:
@ssunday 
I am thinking about placing something along the lines of this in board:
```ruby
def list(fn)
  fn.call(cells.map(:&symbol))
end
```

And then having game be able to retrieve the board representation in the format(s) it wants. I am also considering using blocks. I am attempting to invert control and have dependent classes (game) of board be able to issue a command so that they do not have to adhere to a specific format returned by the board class.

What are your thoughts on this approach?

ssunday replied:
Hmm. I was just thinking have whatever board class return what it knows as its list (like rows, cols, diag)...

christophermanahan replied:
I think that is what I will do for now, but do you think in the future that inverting control (not necessarily in this manner) would be a good design decision in a future refactor? I have been mildly obsessed with trying to think of a way to manage the dependency of game on board.

ssunday replied:
Honestly, I don't have a strong opinion on it. I tend to talk the line of simple/clear and try not to get too fancy with things as it decreases readability/understanding

christophermanahan replied:
Fair enough. Thank you!

ssunday commented on spec/game_spec.rb:
```diff
@@ -9,7 +9,19 @@ def initialize(moves_until_full)
 
   def rows
-    "[#{latest_move.join(' ')}]"
+    if moves_until_full == 0
```
.zero?

christophermanahan committed:
Refactor board and game to use combinations
Privatize diagonals and columns and expose combinations for simpler win checking in game

christophermanahan committed:
Remove unused full method from board

christophermanahan committed:
Remove board rows and refactor game

damonkelley commented on lib/game.rb:
```diff
@@ -6,14 +6,25 @@ def initialize(board:, symbols:, user_interface:)
+      combination.any? do |in_a_row|
+        unique = in_a_row.uniq
+        !unique.first.nil? && unique.size == 1
+      end
```
This could be simpler if the combinations were flattened maybe.

ssunday commented on lib/board.rb:
```diff
@@ -23,7 +17,35 @@ def put(symbol, position)
+  end
+
+  def columns
+    rows.map.with_index do |_, i|
```
😁

christophermanahan committed:
Fix previously incorrect diagonals method

christophermanahan committed:
Refactor board to return flattened combinations
